File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        nodes = []
        # should be [2, 3, 4, 5]
        def traversal(node):
            if not node:
                return

            traversal(node.left)
            nodes.append(node.val)
            traversal(node.right)
            
        
        traversal(root)
        return nodes[k-1]

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> res;
    unordered_set<int> levels;

    void dfs(TreeNode* node, int level) {
        if (levels.find(level) == levels.end()) {
            res.push_back(node->val);
            levels.insert(level);
        } 
        if (node->right != nullptr) dfs(node->right, level+1);
        if (node->left != nullptr) dfs(node->left, level+1);
    }

    vector<int> rightSideView(TreeNode* root) {
        res = {};
        levels = {};
        if (root==nullptr) return res;
        dfs(root, 0);
        return res;
    }
};
